Add a foreign layer URL host whitelist

nwt opened the pull request:
Until we have some experience hosting foreign layer manifests, the Hub
operators wish to limit foreign layers on Hub to those hosted by
Microsoft. To that end, this change adds a registry configuration option
to whitelist hosts that may appear in foreign layer URLs of pushed
manifests, and it modifies manifest validation to reject pushes
containing foreign layer URLs whose host is not whitelisted.

Signed-off-by: Noah Treuhaft <noah.treuhaft@docker.com>

dmp42 commented:
cc @runcom @docker/core-distribution-maintainers 

and related to #1825

runcom commented:
seems inline with #1825 so :+1:

RichardScothern commented on docs/configuration.md:
```diff
@@ -491,6 +495,27 @@ map.
+    foreignlayer:
+        hostwhitelist: [EXAMPLE.COM, .example.net]
+
+Set `hostwhitelist` to an empty list (`[]`) to reject all foreign-layer
```
could you just add an `enabled: <bool>` under the `foreignlayer` section instead?  (more explicit and easier for testing)

nwt replied:
Certainly. And in that case, panicking on an empty whitelist seems appropriate.

stevvooe replied:
Let's use the term `allowed`.

stevvooe replied:
Or better yet:

```
foreignlayer:
  hosts:
    allow:
      - www.foo.com
      - example.com
```

nwt replied:
Happy to follow either this proposal or @RichardScothern's proposal  of `whitelist`.

RichardScothern commented on docs/configuration.md:
```diff
@@ -401,6 +403,8 @@ Permitted values are `error`, `warn`, `info` and `debug`. The default is
       cache:
         blobdescriptor: inmemory
+      foreignlayer:
+        hostwhitelist: [example.com, example.net]
```
Maybe just `whitelist` since there are domains?

nwt replied:
Will do.

RichardScothern commented on registry/handlers/app.go:
```diff
@@ -194,6 +195,28 @@ func NewApp(ctx context.Context, config *configuration.Configuration) *App {
+	if flConfig, ok := config.Storage["foreignlayers"]; ok {
+		if hosts, ok := flConfig["hostwhitelist"]; ok {
+			switch hosts := hosts.(type) {
+			case []interface{}:
```
Is this usable with environment variables? e.g.
REGISTRY_STORAGE_FOREIGNLAYER_HOSTWHITELIST="[a.com]"

aaronlehmann replied:
With the current code I don't think this would work, but you could probably work around it with something like `REGISTRY_STORAGE_FOREIGNLAYER="hostwhitelist: [a.com]"`.

We don't have a good general solution for expressing array values in environment variables.

nwt replied:
Is this good as is or should `REGISTRY_STORAGE_FOREIGNLAYER_HOSTWHITELIST='[a.com]` be made to work?

aaronlehmann replied:
IMHO supporting that variation is out of scope. We should test to ensure something like `REGISTRY_STORAGE_FOREIGNLAYER="hostwhitelist: [a.com]"` is possible, though.

RichardScothern replied:
Yeah, I'm not sure if what you will get will be a `[]interface`, if it comes from an environment variable is may be `[]string`

aaronlehmann replied:
Environment variables are always fed through the YAML parser, so I don't think this is an issue. Again, it's worth testing this.

RichardScothern commented on docs/configuration.md:
```diff
@@ -491,6 +495,27 @@ map.
+match any subdomain while host entries must match exactly.
+
+The following example allows `example.com` and `WWW.EXAMPLE.NET` but not
+`WWW.EXAMPLE.COM` or `www.example.net`.
```
example.net should be accepted.

nwt replied:
Agreed.

stevvooe replied:
I don't understand these limitations. Do we need to employ https://godoc.org/golang.org/x/net/publicsuffix? What are you trying to do?

nwt replied:
I'm trying to separate the whitelist-a-single-name operation from the whitelist-all-subdomains-of-a-name operation.

stevvooe replied:
I see. Leading `.` is the wildcard. Are there other systems that follow this convention?

nwt replied:
http://httpd.apache.org/docs/current/mod/mod_proxy.html#noproxy

RichardScothern commented on registry/handlers/app.go:
```diff
@@ -194,6 +195,28 @@ func NewApp(ctx context.Context, config *configuration.Configuration) *App {
+				for i, host := range hosts {
+					switch host := host.(type) {
+					case string:
+						hostStrings[i] = host
```
Check for valid host/domain format?

nwt replied:
Happy to add that if it's wanted.

stevvooe commented on configuration/configuration.go:
```diff
@@ -350,6 +350,8 @@ func (storage Storage) Type() string {
 			// allow configuration of caching
 		case "delete":
 			// allow configuration of delete
+		case "foreignlayers":
```
What does this have to do with storage? This should be a registry option.

stevvooe replied:
I see why this is a logical place. In the past, we had `Compatibility` to include manifest validation, which was top-level.

stevvooe commented on docs/configuration.md:
```diff
@@ -491,6 +495,27 @@ map.
 >are equivalent, `layerinfo` has been deprecated, in favor or
 >`blobdescriptor`.
 
+### foreignlayers
```
This shouldn't really be called "Foreign Layers". The registry doesn't and shouldn't know what a foreign layer is.

This feature influences the validation of manifest urls.

stevvooe commented:
This feature shouldn't be called "Foreign Layers". The concept of a foreign layer is reserved to the docker engine, as the layer is foreign to docker. The registry doesn't care. This just influences the valid urls that a registry will accept in the manifest.

runcom commented:
Can we take care of enabling/disabling the whole foreign layer story/feature as part of this PR? Related to poing 3) in https://github.com/docker/distribution/issues/1825

Is it feasible in this PR? seems likely to me, just an `enabled` flag under `foreignLayers:` opt (or whatever it'll be called)
If not, I'll take care of it as a follow up (but I don't see issues in doing this here.

The point from https://github.com/docker/distribution/issues/1825 is:
> 3) provide a way to opt-in for such a feature at the registry level
Since there may be the need to defer this support on some deployment, we can have an explicit setting which opt-in for this new behavior.

@dmp42 @stevvooe wdyt?

runcom commented on registry/storage/registry.go:
```diff
@@ -154,6 +171,20 @@ func (reg *registry) BlobStatter() distribution.BlobStatter {
 
+// ForeignLayerHostWhitelist tests whether the host is allowed in foreign layer URLs.
+func (reg *registry) ForeignLayerHostAllowed(host string) bool {
+	if reg.foreignLayerHostWhitelist == nil {
```
the knob I'm talking about can cleanly take care of this check also - instead of checking for the presence of a whitelist, we can check if foreign layers are enabled at all and exit early here

nwt commented:
> This feature shouldn't be called "Foreign Layers". 

I'm happy follow maintainer guidance on an appropriate name. I don't think I have enough perspective to propose an alternative. Except for "Not Foreign Layers". 💥

nwt commented:
> Is it feasible in this PR? seems likely to me, just an `enabled` flag under `foreignLayers:` opt (or whatever it'll be called)

@RichardScothern proposed [the same](https://github.com/docker/distribution/pull/1829#discussion_r70340968) and I'm planning to implement it.

runcom commented:
> Is it feasible in this PR? seems likely to me, just an enabled flag under foreignLayers: opt (or whatever it'll be called)
> @RichardScothern proposed the same and I'm planning to implement it.

ty

stevvooe commented:
@nwt This is validating manifest urls:

```
validation:
  disabled: false
  manifests:
    urls:
      allowed:
        - https://foo.com
```

A few things:

1. This is configuring how we validate incoming manifests. While this is on storage configuration, let's consider pulling the validation above storage.
2. We should also think about prevention of serving manifests with disallowed urls. For example, if it doesn't pass url validation, should we serve it at all? This is important since we don't block these _at all_ now.
3. We should be validating URLs, not hosts and domains.

nwt commented:
@stevvooe: I implemented this:
```
validation:
  manifests:
    urls:
      disabled: false
      allowed: regexp
```
I changed `allowed` to a regular expression because Hub needs more control than a list of strings can provide, and I moved `disabled` under `urls` since that's the only validation it influences.

runcom commented:
I meant to have disabled (or enabled) as a way to fully opt-out this features. It seems what you did doesn't cleanly achieve that, does it?

nwt commented:
@runcom: I think it does. When

nwt commented:
@runcom: All pushes of foreign layer manifests will be rejected with this in your config:
```
validation:
  manifests:
    urls:
      disabled: true
```

Were you asking for something different?

runcom commented:
That's somehow fine ti me, i was hoping for such a flag to be more global but if it still accomplishes fully opt-out than I'll leave those decisions to others about where that disable flag should be.

stevvooe commented on configuration/configuration.go:
```diff
@@ -172,6 +172,15 @@ type Configuration struct {
+	Validation struct {
+		Manifests struct {
+			URLs struct {
+				Allowed  *string `yaml:"allowed,omitempty"`
```
Why not a list?

stevvooe commented on configuration/configuration.go:
```diff
@@ -172,6 +172,15 @@ type Configuration struct {
 		} `yaml:"schema1,omitempty"`
 	} `yaml:"compatibility,omitempty"`
+
+	Validation struct {
```
Not sure how this passes the linter, but please document the field.

aaronlehmann replied:
I believe the linter doesn't care about documentation for nested structs, only structs defined at the top level.

stevvooe commented on docs/configuration.md:
```diff
@@ -1796,6 +1801,51 @@ defines a such a feature with configurable behavior.
+
+Use the `manifest` subsection to configure manifest validation.
+
+#### URLs
```
Make this clear that this applies to any descriptors that have urls in a manifest. This validation needs to include any descriptor in manifest list, as well.

stevvooe commented on registry/storage/registry.go:
```diff
@@ -46,6 +52,23 @@ func DisableDigestResumption(registry *registry) error {
 
+// DisableManifestURLs is a functional option for NewRegistry. It causes the
+// registry to reject pushed manifests containing URLs.
+func DisableManifestURLs(registry *registry) error {
```
O, these are separate features? Why not just have a `deny` section?

```
validation:
  enabled: true # enabled turns on the feature
  manifests:
    urls:
      deny:
        - *
```

Typically, we use `enabled`/`disabled` as a feature switch. Having mean that "urls are disabled" is confusing.

runcom commented:
@stevvooe +1

nwt commented:
@stevvooe: 153c2f2cb946839a10837d771111658a89e9234d should be closer to what you had in mind than 2e376295a135821963afd7c6eae2c200b30dbbb6 was, with `enabled` immediately below `validation`, `deny` as well as `allow`, and lists for both.

stevvooe commented on docs/configuration.md:
```diff
@@ -1796,6 +1802,39 @@ defines a such a feature with configurable behavior.
+      disabled: false
+      manifests:
+        urls:
+          allow: ^https?://([^/]+\.)*example\.com/
```
I think these examples need to be a list.

stevvooe commented on configuration/configuration.go:
```diff
@@ -172,6 +172,24 @@ type Configuration struct {
+	// Validation configures validation options for the registry.
+	Validation struct {
+		// Disabled disables all other options in this section.
+		Disabled bool `yaml:"disabled,omitempty"`
```
By default, this should be "disabled". The zero-value of `Disabled`, would be `false`, which would mean `Enabled`. This field should be called `Enabled` and users should be required to set it to true to enable the validation.

I think, by default, we can refuse manifests with anything in the url field.

stevvooe commented:
LGTM

There are a few nits about getting the defaults backwards compatible.

Great addition, @nwt!

nwt commented:
@stevvooe: 61e5803b56a41292779bdd259ea14f166531579e addresses comments on 153c2f2cb946839a10837d771111658a89e9234d.

1. Documentation examples properly show `allow` and `deny` as lists.
1. `Configuration.Validation.Enabled` replaces `….Disabled`.
1. If `allow` and `deny` are empty, all manifests with URLs are rejected.


I'm not thrilled with how I handled for that last one, both because`deny` is no longer useful without `allow` and because `storage.ManifestURLsAllowRegexp(regexp.MustCompile("^$"))` in app.go is a bit magical, so let me know if you think either of those edges needs filing.

stevvooe commented:
@nwt 

> I'm not thrilled with how I handled for that last one, both becausedeny is no longer useful without allow and because `storage.ManifestURLsAllowRegexp(regexp.MustCompile("^$"))` in app.go is a bit magical, so let me know if you think either of those edges needs filing.

Will it still allow manifests _without_ urls? That is somewhat the purpose. For example, if one configures the following:

```
validation:
  enabled: true
```

All manifests with urls would be blocked, while others would be allowed.

Let's say we unrolled these into an ordered set of operations. By default, with all empy `allow` and `deny`, it would look like this:

```
DENY ALL
```

This would give us default closed operation when the URL is empty.

We insert all allows before denies:

```
ALLOW foo.com
DENY ALL
```

ALLOW always has precedent:

```
ALLOW foo.com
DENY foo.com
DENY ALL
```

These are matched such that if the rule matches, we return the operation. For `foo.com`, we'd match an `ALLOW` rule first and the URL would be allowed.

This only becomes problematic when we have a more specific deny than an allow:

```
ALLOW .*
DENY foo.com
DENY ALL
```

So, even though we deny `foo.com`, allow has precedent and is match before the deny. We can fix this by ordering DENY before allow, but we get the same issue, except we have allow that is more specific than a deny:

```
DENY .*
ALLOW foo.com
DENY ALL
```

Typically, this issue is dealt with using priorities, but that is overbuilt for this purpose. It is probably better to favor ALLOW for this purpose but arguments can go either way.

__TL; DR__ Having `^$` as a "magic" deny entry is an okay bit of magic. That defines the default operation on the match chain.

nwt commented:
@stevvooe:

> Will it still allow manifests without urls?

Yes, manifests without URLs are always allowed regardless of what appears under `validation`.

RichardScothern commented:
LGTM

stevvooe commented:
LGTM

dmcgowan commented:
LGTM

RichardScothern commented:
Holding off on merging while we discuss a missing API test..

RichardScothern commented:
Fixes a pre-existing condition and coming in a PR.
